multy_core/src/bitcoin/bitcoin_transaction_segwit.cpp

```diff
@@ -0,0 +1,506 @@
+    {
+    }
+
+    const BigInt& get_amount_per_byte() const
+    {
+        return amount_per_byte.get_value();
+    }
+
+    void validate_fee(const BigInt& leftover, uint64_t transaction_size) const
+    {
+        const BigInt min_transaction_fee(amount_per_byte.get_value() * transaction_size);
+
+        if (leftover < min_transaction_fee)
+        {
+            THROW_EXCEPTION2(ERROR_TRANSACTION_FEE_TOO_LOW,
+                    "Transaction total fee is too low. ")
+                    << leftover.get_value() << " < " << min_transaction_fee.get_value();
+        }
+    }
+
+public:
+    PropertyT<BigInt> amount_per_byte;
+};
+
+BitcoinTransactionSegWit::BitcoinTransactionSegWit(BlockchainType blockchain_type)
+    : TransactionBase(blockchain_type),
+      m_version(2),
+      m_fee(new BitcoinTransactionSegWitFee),
+      m_sources(),
+      m_destinations(),
+      lock_time(make_clone(BinaryData{BITCOIN_LOCK_TIME, 4}),
+                    m_properties, "nlocktime", Property::OPTIONAL)
+{
+    register_properties("", m_fee->get_properties());
+}
+
+Properties& BitcoinTransactionSegWit::add_source()
+{
+    BitcoinTransactionSegWitSourcePtr source(new BitcoinTransactionSegWitSource());
+    m_sources.emplace_back(std::move(source));
+
+    return register_properties(
+            make_id("#", m_sources.size() - 1),
+            m_sources.back()->get_properties());
+}
+
+Properties& BitcoinTransactionSegWit::add_destination()
+{
+    BitcoinTransactionSegWitDestinationPtr destination(
+            new BitcoinTransactionSegWitDestination(get_net_type()));
+    m_destinations.emplace_back(std::move(destination));
+
+    return register_properties(
+            make_id("#", m_destinations.size() - 1),
+            m_destinations.back()->get_properties());
+}
+
+Properties& BitcoinTransactionSegWit::get_fee()
+{
+    return m_fee->get_properties();
+}
+
+void BitcoinTransactionSegWit::set_message(const BinaryData& /*value*/)
+{
+    THROW_EXCEPTION("Not implemented yet");
+}
+
+BigInt BitcoinTransactionSegWit::estimate_total_fee(size_t /*sources_count*/, size_t /*destinations_count*/) const
+{
+    THROW_EXCEPTION("Not implemented yet");
+}
+
+BigInt BitcoinTransactionSegWit::get_total_fee() const
+{
+    return calculate_diff();
+}
+
+BigInt BitcoinTransactionSegWit::get_total_spent() const
+{
+    THROW_EXCEPTION("Not implemented yet");
+}
+
+void BitcoinTransactionSegWit::update()
+{
+    verify();
+    sign();
+
+}
+
+BinaryDataPtr BitcoinTransactionSegWit::serialize()
+{
+    update();
+
+    BitcoinDataStream data_stream;
+    data_stream << m_version;
+    data_stream << static_cast<uint8_t>(0x00); // marker
+    data_stream << static_cast<uint8_t>(0x01); // flag
+    data_stream << as_compact_size(m_sources.size());
+    for (const auto& source : m_sources)
+    {
+        data_stream << *source->serialize_with_script_pub_key();
+    }
+    data_stream << as_compact_size(m_destinations.size());
```

Comment:
You are serializing empty destinations here to, needlessly increasing transaction size.